### PR TITLE
[DQL-188] EPIC - Dataset comment enhancements

### PR DESCRIPTION
**Pull request by salsa-nathan**
- [DQL2-28] Added title and comment to new comment email template
- [DQL2-28] Updated notification-new-comment email body template with update variable names
- [DQL2-28] Updated dataset_title to more generic variable name
- [DQL2-28] Removed unused method notify_admins
- [DQL2-28] Updated variable name content_title to content_item_title
- [DQL2-26] Record which user deletes a comment in the DB
- [DQL2-26] Added UpdateDBCommand paster command for DB updates
- [DQL2-26] Updated readme with initdb & updatedb paster commands
- [DQL2-25] Added configurable option to display dataset comments on a tab page instead of below dataset additional information
- [DQL2-25] Reversed logic for better readability
- [DQL2-24] Added comments count badge and helpers
- [DQL2-24] Refactored plugin methods into helpers
- [DQL2-25] Fixed redirect bug for datarequest comments
- [DQL2-25] Handle email notifications for tabbed page in `get_content_item_link`.
- [DQL2-26] Added snippet for deleted comment message.
- [DQL2-23] Added new BDD tests for:
  - Comment tab
  - Comment count badge
  - Deleted by username
- [DQL2-23] Bump CKAN version to 2.8.4
- [DQL2-23] Updated method name to match step name
Formated code
- [DQL2-28] Formatted files util.py, email_notifications.py and marked method notify_admins as deprecated
- [DQL2-24] Formatted file count_badge.html, controller.py, helpers.py
- [DQL2-26] Formatted file command.py

### CODE REVIEW

File: test/features/steps/steps.py [hunk 1]
```diff
@@ -30,6 +30,14 @@ def go_to_dataset(context, name):
     when_i_visit_url(context, '/dataset/' + name)
 
 
+@step(u'I go to dataset "{name}" comments')
+def go_to_data_request_comments(context, name):
```
Comment (ThrawnCA) [inline test/features/steps/steps.py:34]: Misleading function name.

File: ckanext/ytp/comments/email_notifications.py [hunk 2]
```diff
@@ -151,30 +152,7 @@ def get_admins(owner_org, user, content_type, content_item_id):
     return users
 
 
-def notify_admins(owner_org, user, template, content_type, content_item_id, comment_id):
-    """
-
-    :param owner_org: organization.id of the content item owner
```
Comment (duttonw) [inline ckanext/ytp/comments/email_notifications.py:157]: why comment deleted?

File: ckanext/ytp/comments/email_notifications.py [hunk 3]
```diff
@@ -209,7 +187,9 @@ def notify_admins_and_comment_notification_recipients(owner_org, user, template,
             users,
             template,
             {
-                'url': get_content_item_link(content_type, content_item_id, comment_id)
+                'url': get_content_item_link(content_type, content_item_id, comment_id),
+                'content_item_title': content_item_title,
```
Comment (duttonw) [inline ckanext/ytp/comments/email_notifications.py:191]: why not remove html markup from item title, better safe their sorry.

File: ckanext/ytp/comments/templates/snippets/count_badge.html [hunk 4]
```diff
@@ -0,0 +1 @@
+<span class="badge">{{ count }}</span>
```
Comment (duttonw) [inline ckanext/ytp/comments/templates/snippets/count_badge.html:1]: should we have a new line at the end of the file?

File: ckanext/ytp/comments/util.py [hunk 5]
```diff
@@ -27,3 +28,15 @@ def clean_input(comment):
...
+    try:
+        # Returns the text content of the element, including the text content of its children, with no HTML markup.
+        return fromstring(text.replace('<br/>', '\n')).text_content()
+    except Exception, e:
+        if type(e).__name__ == "ParserError":
+            raise logic.ValidationError("Text is required")
+        else:
+            template = "An exception of type {0} occurred. Arguments:\n{1!r}"
+            message = template.format(type(e).__name__, e.args)
+            log.debug(message)
```
Comment (duttonw) [inline ckanext/ytp/comments/util.py:42]: should we have a new line at the end of the file?

File: requirements-dev.txt [hunk 6]
```diff
@@ -5,8 +5,8 @@ flake8==3.7.9
 nose==1.3.7
 splinter>=0.13.0
 -e git+https://github.com/qld-gov-au/ckanext-datarequests@training#egg=ckanext-datarequests
--e git+https://github.com/qld-gov-au/ckanext-data-qld@develop#egg=ckanext-data_qld
+-e git+https://github.com/salsadigitalauorg/ckanext-data-qld@develop#egg=ckanext-data_qld
```
Comment (duttonw) [inline requirements-dev.txt:8]: todo, revert back to qld-gov-au once merged.

File: requirements-dev.txt [hunk 7]
```diff
@@ -5,8 +5,8 @@ flake8==3.7.9
...
 -e git+https://github.com/ckan/ckantoolkit@release-0.0.4#egg=ckantoolkit
 -e git+https://github.com/ckan/ckanapi@ckanapi-4.3#egg=ckanapi
 -e git+https://github.com/ckan/ckanext-scheming@release-1.2.0#egg=ckanext-scheming
--e git+https://github.com/qld-gov-au/ckanext-data-qld-theme@develop#egg=ckanext-data_qld_theme
+-e git+https://github.com/salsadigitalauorg/ckanext-data-qld-theme@develop#egg=ckanext-data_qld_theme
```
Comment (duttonw) [inline requirements-dev.txt:12]: todo, revert back to qld-gov-au once merged.

File: ckanext/ytp/comments/email_notifications.py (hunk 2 above)
Comment (MarkCalvert) [reply ckanext/ytp/comments/email_notifications.py:157]: Hi @duttonw, while implementing this story https://salsadigital.atlassian.net/browse/DQL2-28, the solution direction mentioned "Adjust the notify_admins and notify_admins_and_comment_notification_recipients functions in helpers.py of the ckanext-ytp-comments extension" however I noticed the method 'notify_admins' was not referenced anywhere in the code, so I decided to remove the method. Is there a reason you want me to re-add it?

Comment (duttonw) [reply ckanext/ytp/comments/email_notifications.py:157]: is good to prune unused code. 👍

Comment (duttonw) [reply ckanext/ytp/comments/email_notifications.py:157]: actually thinking about it, we should leave this in and mark it as deprecated as others may be referencing this as this is a plugin that more than us use. (if we want to contribute back to the main repo)

Comment (MarkCalvert) [reply ckanext/ytp/comments/email_notifications.py:157]: Yeh that could be true, will add it back in.

File: test/features/steps/steps.py (hunk 1 above)
Comment (MarkCalvert) [reply test/features/steps/steps.py]: @ThrawnCA Updated the method name to match step name

File: ckanext/ytp/comments/email_notifications.py (hunk 2 above)
Comment (MarkCalvert) [reply ckanext/ytp/comments/email_notifications.py:157]: @duttonw Re-added 'notify_admins' method with a deprecation decorator

File: ckanext/ytp/comments/email_notifications.py (hunk 3 above)
Comment (MarkCalvert) [reply ckanext/ytp/comments/email_notifications.py]: @duttonw removed html markup from item title to play it safe

File: ckanext/ytp/comments/templates/snippets/count_badge.html (hunk 4 above)
Comment (MarkCalvert) [reply ckanext/ytp/comments/templates/snippets/count_badge.html]: @duttonw Added new line to end of the file

File: ckanext/ytp/comments/util.py (hunk 5 above)
Comment (MarkCalvert) [reply ckanext/ytp/comments/util.py]: @duttonw Added new line to end of the file

**Comment by MarkCalvert** [comment]
Hi @duttonw & @ThrawnCA,

I have committed some code changes with the suggestions you guys mentioned.
Over to you guys to review and resolve.

Cheers

File: test/features/comments.feature [hunk 8]
```diff
@@ -79,29 +79,102 @@ Feature: Comments
...
+        When I log in
+        And I go to data request "Test Request" comments
+        And I press the element with xpath "//a[@title='Delete comment']"
+        Then I should see "Are you sure you want to delete this comment?" within 1 seconds
+        Then I press the element with xpath "//button[contains(string(), 'Confirm')]"
+        Then I take a screenshot
+        Then I should not see "This comment was deleted." within 2 seconds
+
+    @comment-add
+    Scenario: When a logged-in user submits a comment on a Dataset the page should be scrolled to the highlighted comment I just posted.
```
Comment (ThrawnCA) [inline test/features/comments.feature:110]: This sentence switches between third-person and first-person.

File: test/features/comments.feature [hunk 9]
```diff
@@ -79,29 +79,102 @@ Feature: Comments
...
         And I press the element with xpath "//a[@title='Delete comment']"
         Then I should see "Are you sure you want to delete this comment?" within 1 seconds
         Then I press the element with xpath "//button[contains(string(), 'Confirm')]"
         Then I take a screenshot
-        Then I should see "This comment was deleted." within 2 seconds
+        Then I should see "This comment was deleted by admin." within 2 seconds
+
+    @comment-tab
+    Scenario: Non-logged in users should not see comment form in dataset tab
+        Given "Unathenticated" as the persona
```
Comment (ThrawnCA) [inline test/features/comments.feature:156]: Misspelled "Unauthenticated". Possibly this is misspelled in the original definition?

File: test/features/comments.feature (hunk 8 above)
Comment (salsa-nathan) [reply test/features/comments.feature]: @ThrawnCA have addressed via https://github.com/MarkCalvert/ckanext-ytp-comments/commit/036112d3f95d8b4f99b9471c3c6c0d42903dd04c

File: test/features/comments.feature (hunk 9 above)
Comment (salsa-nathan) [reply test/features/comments.feature]: @ThrawnCA have addressed via https://github.com/MarkCalvert/ckanext-ytp-comments/commit/036112d3f95d8b4f99b9471c3c6c0d42903dd04c

File: requirements-dev.txt (hunk 6 above)
Comment (salsa-nathan) [reply requirements-dev.txt]: @duttonw - I have updated this via commit https://github.com/MarkCalvert/ckanext-ytp-comments/commit/036112d3f95d8b4f99b9471c3c6c0d42903dd04c

File: requirements-dev.txt (hunk 7 above)
Comment (duttonw) [reply requirements-dev.txt:12]: is it possible to run this without our theme as i don't think ckanext-ytp-comments wants to be dependant on our theme for their success.

Comment (salsa-nathan) [reply requirements-dev.txt:12]: @duttonw Hmm, yes I hear you re: making the `ckanext-ytp-comments` tests independent of the data-qld-theme extension, however it's a little bit of a grey area at the moment. The ability to display comments on a separate page/tab to the dataset details is now configurable, so the decision to display them on a separate tab would occur on a per-site basis. In this case, it is a requirement for the Data.Qld site, so the template for doing so has been implemented within the `ckanext-data-qld-theme` extension, and so the tests are based on the assumption that `show_comments_tab_page` config setting is set to True. I will try to think of a way we could incorporate this and keep it as generic as possible while keeping in mind that we've already demolished the timebox provision for automated testing on this epic... can we proceed with getting this one into QA in the meantime, or does this need to be resolved immediately? Cheers.

Comment (duttonw) [reply requirements-dev.txt:12]: lets add a ticket to the backlog and move forward then.

Comment (duttonw) [reply requirements-dev.txt:12]: have merged the theme, please revert back to qld-gov-au version please.

**Comment by salsa-nathan** [comment]
Hi @duttonw & @ThrawnCA 
I have reverted the `ckanext-data-qld-theme` extension to use the `qld-gov-au` repo now.
Thanks

**Pull request merged**